gdal/swig/python/gdal-utils/osgeo_utils/auxiliary/gdal_argparse.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+# ******************************************************************************
+#
+#  Project:  GDAL utils.auxiliary
+#  Purpose:  Extended argparse.GDALArgumentParser to be used with all gdal-utils
+#  Author:   Idan Miara <idan@miara.com>
+#
+# ******************************************************************************
+#  Copyright (c) 2021, Idan Miara <idan@miara.com>
+#
+#  Permission is hereby granted, free of charge, to any person obtaining a
+#  copy of this software and associated documentation files (the "Software"),
+#  to deal in the Software without restriction, including without limitation
+#  the rights to use, copy, modify, merge, publish, distribute, sublicense,
+#  and/or sell copies of the Software, and to permit persons to whom the
+#  Software is furnished to do so, subject to the following conditions:
+#
+#  The above copyright notice and this permission notice shall be included
+#  in all copies or substantial portions of the Software.
+#
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+#  OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+#  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+#  DEALINGS IN THE SOFTWARE.
+# ******************************************************************************
+
+import sys
+import argparse
+import shlex
+
+
+class ExtendAction(argparse.Action):
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        items = getattr(namespace, self.dest) or []
+        items.extend(values)
+        setattr(namespace, self.dest, items)
+
+
+class GDALArgumentParser(argparse.ArgumentParser):
+
+    def __init__(self, fromfile_prefix_chars='@', **kwargs):
+        super().__init__(fromfile_prefix_chars=fromfile_prefix_chars, **kwargs)
+        if sys.version_info < (3, 8):
+            # extend was introduced to the stdlib in Python 3.8
+            self.register('action', 'extend', ExtendAction)
+
+    def parse_args(self, args=None, optfile_arg=None, **kwargs):
+        if (args is not None
+           and optfile_arg in args
+           and self.fromfile_prefix_chars is not None
+           and optfile_arg is not None):
+
+            # Replace '--optfile x' with the standard '@x'
```

Comment:
I'm not sure they are completely equivalent. From what I can read in https://docs.python.org/3/library/argparse.html#fromfile-prefix-chars , @x requires one argument per line whereas --optfile x can support potentially arguments on the same line

Reply:
Thanks for your comment.
I actually handled that in GDALArgumentaParser. 
It passes the relevant test7 for that cases you've mentioned.

Reply:
ok, any reason to have optfile_arg default to None and being set only in gdal_calc.py to "--optfile" ?

And is this code exercised by the test suite ? (my mantra is "non-tested code is broken, or will soon be broken", and it is even more true for an interpreted language like Python)

Reply:
--optfile is for backwards compatibility. I can't see a reason to encourage people to use it ovet the standard pefix@.
I'll set it on any until that currently have this option.

I've learned to know your mantra :) 
Almost all the tests of `gdal_calc` use the ArgParser so it is tested there through it.
Did you refer to any other code?

Reply:
Regarding testing, I refer to this new code replacing --optfile by ``@ ``

Reply:
Calc test 7 tests that as it inputs `- -optfile` and the parser transforms it to `@` and parses it as the normal way. 

I can add a test for input with the prefix.

Reply:
I've added tests for `@` prefix and a deprecation warning for `--optfile`.